### PR TITLE
add cython bindings

### PR DESCRIPTION
**Pull request by arupcsedu**

### CODE REVIEW

**Comment by nirandaperera** [comment]
@arupcsedu I'm closing this PR because the branch commit history is way too jumbled . Please create a branch from cylon:main and cherrypick your latest commits to it and send a new PR. And please don't submit PRs from your main branch. Always keep it synced with cylon:main and do your new work on a separate branch.

File: cpp/src/cylon/indexing/slice.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  const auto &in_table = in->get_table();
+
+  std::shared_ptr<arrow::Table> out_table;
+  if (!in->Empty()) {
+    out_table = in_table->Slice(offset, length);
+  } else {
+    out_table = in_table;
+  }
+  return Table::FromArrowTable(ctx, std::move(out_table), *out);
+}
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:45]: There is already a `out` pointer and also there is another `out_table` `shared_ptr` reference created to extract output. Plus the method takes in a Table `shared_ptr` and a pointer. This looks kind of redundant and not clean as per my understanding. May be the required change to the method is out of scope for this PR. 

@nirandaperera appreciate your feedback on this method and usage.

File: cpp/src/cylon/indexing/slice.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            ->raw_values());
+  }
+
+  int64_t rank = ctx->GetRank();
+  int64_t prefix_sum = std::accumulate(partition_lengths, partition_lengths + rank, kZero);
+  int64_t total_length = std::accumulate(partition_lengths + rank,
+                                         partition_lengths + ctx->GetWorldSize(),
+                                         prefix_sum);
+  if (global_offset > total_length) {
+    return {Code::Invalid, "global offset exceeds total length of the dist table"};
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:76]:
```suggestion
    return {Code::Invalid, "global offset exceeds total length of the distributed table"};
```

File: cpp/src/cylon/indexing/slice.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  }
+  // adjust global length if it exceeds total_length
+  if (global_offset + global_length > total_length) {
+    global_length = total_length - global_offset;
+  }
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:81]: cc @nirandaperera is this okay?

File: cpp/src/cylon/indexing/slice.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  int64_t local_length =
+      std::min(this_length, std::max(global_offset + global_length - prefix_sum, kZero))
+          - local_offset;
+
+  return Slice(in, local_offset, local_length, out);
+}
+
+Status DistributedSlice(const std::shared_ptr<Table> &in,
+                        int64_t offset,
+                        int64_t length,
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:96]: should it be?
```suggestion
                        const int64_t offset,
                        const int64_t length,
```

File: cpp/src/cylon/indexing/slice.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+ */
+Status Head(const std::shared_ptr<Table> &table, int64_t num_rows,
+            std::shared_ptr<Table> *output) {
+  if (num_rows >= 0) {
+    return Slice(table, 0, num_rows, output);
+  } else
+    return {Code::Invalid, "Number of head rows should be >=0"};
+}
+
+Status DistributedHead(const std::shared_ptr<Table> &table, int64_t num_rows,
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:114]: should it be?
```suggestion
Status DistributedHead(const std::shared_ptr<Table> &table, const int64_t num_rows,
```

File: cpp/src/cylon/indexing/slice.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    return {Code::Invalid, "Number of head rows should be >=0"};
+  }
+}
+
+/**
+ * Tail the part of table to create a single table with specific number of rows
+ * @param tables, number of rows
+ * @return new table
+ */
+Status Tail(const std::shared_ptr<Table> &table, int64_t num_rows,
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:131]: should it be?
```suggestion
Status Tail(const std::shared_ptr<Table> &table, const int64_t num_rows,
```

File: cpp/src/cylon/indexing/slice.cpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  const int64_t table_size = in_table->num_rows();
+
+  if (num_rows >= 0) {
+    return Slice(table, table_size - num_rows, num_rows, output);
+  } else {
+    return {Code::Invalid, "Number of tailed rows should be >=0"};
+  }
+}
+
+Status DistributedTail(const std::shared_ptr<Table> &table, int64_t num_rows,
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:144]: should it be?
```suggestion
Status DistributedTail(const std::shared_ptr<Table> &table, const int64_t num_rows,
```

File: cpp/src/cylon/indexing/slice.cpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                       std::shared_ptr<Table> *output) {
+  if (num_rows >= 0) {
+    const auto &ctx = table->GetContext();
+    std::shared_ptr<cylon::Column> partition_len_col;
+    const auto &num_row_scalar = Scalar::Make(arrow::MakeScalar(table->Rows()));
+    RETURN_CYLON_STATUS_IF_FAILED(ctx->GetCommunicator()
+                                      ->Allgather(num_row_scalar, &partition_len_col));
+    assert(ctx->GetWorldSize() == partition_len_col->length());
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:152]: Curious: So we limit the number of partitions to the number of processes? Why is that? Could it be multiples of number of processes?

File: cpp/src/cylon/indexing/slice.cpp [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    auto *partition_lengths =
+        std::static_pointer_cast<arrow::Int64Array>(partition_len_col->data())
+            ->raw_values();
+
+    int64_t dist_length =
```
Comment (vibhatha) [inline cpp/src/cylon/indexing/slice.cpp:157]: may be?
```suggestion
    const int64_t dist_length =
```

File: cpp/src/cylon/table.hpp [hunk 10]
```diff
@@ -481,8 +481,51 @@ Status WriteParquet(const std::shared_ptr<cylon::CylonContext> &ctx,
...
                     const io::config::ParquetOptions &options = cylon::io::config::ParquetOptions());
+
 #endif  // BUILD_CYLON_PARQUET
 
+/**
+   * Slice the part of table to create a single table
+   * @param tables, offset, length
+   * @return new sliced table
+   */
+Status Slice(const std::shared_ptr<Table> &in, int64_t offset, int64_t length,
```
Comment (vibhatha) [inline cpp/src/cylon/table.hpp:492]: @nirandaperera we didn't add `Result` like API right?

File: cpp/src/cylon/table.hpp [hunk 11]
```diff
@@ -481,8 +481,51 @@ Status WriteParquet(const std::shared_ptr<cylon::CylonContext> &ctx,
...
+               std::shared_ptr<cylon::Table> *out);
+
+
+/**
+   * Head the part of table to create a single table with specific number of rows
+   * @param tables, number of rows
+   * @return new  table
+   */
+
+Status Head(const std::shared_ptr<Table> &table, int64_t num_rows,
```
Comment (vibhatha) [inline cpp/src/cylon/table.hpp:512]: maybe?

```suggestion
Status Head(const std::shared_ptr<Table> &table, const int64_t num_rows,
```

File: cpp/src/cylon/table.hpp [hunk 12]
```diff
@@ -481,8 +481,51 @@ Status WriteParquet(const std::shared_ptr<cylon::CylonContext> &ctx,
...
+              std::shared_ptr<cylon::Table> *output);
+Status DistributedHead(const std::shared_ptr<Table> &table, int64_t num_rows,
```
Comment (vibhatha) [inline cpp/src/cylon/table.hpp:514]: maybe?

```suggestion
Status DistributedHead(const std::shared_ptr<Table> &table, const int64_t num_rows,
```

File: cpp/src/cylon/table.hpp [hunk 13]
```diff
@@ -481,8 +481,51 @@ Status WriteParquet(const std::shared_ptr<cylon::CylonContext> &ctx,
...
+              std::shared_ptr<cylon::Table> *output);
+
+/**
+   * Tail the part of table to create a single table with specific number of rows
+   * @param tables, number of rows
+   * @return new table
+   */
+
+Status Tail(const std::shared_ptr<Table> &table, int64_t num_rows,
```
Comment (vibhatha) [inline cpp/src/cylon/table.hpp:523]: maybe? 
```suggestion
Status Tail(const std::shared_ptr<Table> &table, const int64_t num_rows, 
```

**Comment by nirandaperera** [comment]
Closing duplicate #619

File: cpp/src/cylon/thridparty/flat_hash_map/bytell_hash_map.hpp [hunk 14]
```diff
@@ -13,6 +13,7 @@
 #include <type_traits>
 #include <vector>
 #include <array>
+#include <limits>
```
Comment (vibhatha) [inline cpp/src/cylon/thridparty/flat_hash_map/bytell_hash_map.hpp:16]: let's re-order?
```suggestion
#include <array>
#include <limits>
#include <type_traits>
#include <vector>
```

File: cpp/src/examples/join_example.cpp [hunk 15]
```diff
@@ -109,6 +111,7 @@ int main(int argc, char *argv[]) {
             << std::chrono::duration_cast<std::chrono::milliseconds>(
                 join_end_time - read_end_time).count() << "[ms]";
   std::vector<std::string> column_names = joined->ColumnNames();
+
   ctx->Finalize();
   return 0;
```
Comment (vibhatha) [inline cpp/src/examples/join_example.cpp:116]: nit: maybe?
```suggestion
  return EXIT_SUCCESS;
```

File: cpp/src/examples/slice_example.cpp [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+  
+  std::shared_ptr<cylon::CylonContext> ctx;
+  if (!cylon::CylonContext::InitDistributed(mpi_config, &ctx).is_ok()) {
+    std::cerr << "ctx init failed! " << std::endl;
+    return 1;
+  }
+
+
+  std::shared_ptr<cylon::Table> in_table, joined, sliced, head_table, tail_table;
+  auto read_options = cylon::io::config::CSVReadOptions().UseThreads(false).BlockSize(1 << 30);
```
Comment (vibhatha) [inline cpp/src/examples/slice_example.cpp:51]: We should probably set default values to the `BlockSize`function and not pass in it here if it is not necessary.

File: cpp/src/examples/slice_example.cpp [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+    //cylon::FromCSV(ctx, std::string(argv[3]) + std::to_string(ctx->GetRank()) + ".csv", first_table);
+    //cylon::FromCSV(ctx, std::string(argv[4]) + std::to_string(ctx->GetRank()) + ".csv", second_table);
+
+    offset = std::stoull(argv[4]);
+    length = std::stoull(argv[5]);
+  }
+  ctx->Barrier();
+  auto read_end_time = std::chrono::steady_clock::now();
+  //in_table->Print();
```
Comment (vibhatha) [inline cpp/src/examples/slice_example.cpp:84]: remove?

File: python/pycylon/pycylon/data/table.pyx [hunk 18]
```diff
@@ -1260,10 +1310,30 @@ cdef class Table:
...
+                new_tb = self.from_arrow(self.context, py_arrow_table.slice(key.start, key.stop))
+                new_index = tb_index.values[key.start: key.stop].tolist()
+                new_tb.set_index(new_index)
+                return new_tb
+            elif key.start == None and key.stop > 0:
+                return self._get_head_response(0, key.stop)
+            elif key.start == None and key.stop < 0:
+                    return self._get_tail_response(0, key.stop*(-1))
+            else:
+                raise ValueError(f"Unsupported Key Type in __getitem__ {type(key)}")
```
Comment (vibhatha) [inline python/pycylon/pycylon/data/table.pyx:1323]: Thanks for adding this. This was something missing in the initial implementation.

**Review by vibhatha — CHANGES_REQUESTED** [review]
I added a few suggestions and thanks a lot for working on this. 
Probably my comments on `const` could be wrong. But please verify it. 

I still have to do a pass on logics in a deeper level. I mainly focused on the grammar and basic logics.

File: python/pycylon/pycylon/data/table.pyx (hunk 18 above)
Comment (arupcsedu) [reply python/pycylon/pycylon/data/table.pyx:1323]: @vibhatha 
This is a wrong pull request. Niranda is supposed to close this.

Please review this one : https://github.com/cylondata/cylon/pull/619

Comment (nirandaperera) [reply python/pycylon/pycylon/data/table.pyx:1323]: Yep. My bad! Sorry @arupcsedu and @vibhatha.

**Pull request closed without merging**